bgpd/bgp_nb_config.c

```diff
@@ -33,12 +78,75 @@
 int routing_control_plane_protocols_control_plane_protocol_bgp_global_create(
 	struct nb_cb_create_args *args)
 {
+
+	const struct lyd_node *vrf_dnode;
+	struct bgp *bgp;
+	const char *vrf_name;
+	const char *name = NULL;
+	as_t as;
+	enum bgp_instance_type inst_type;
+	bool is_view_inst = false;
+	int ret;
+	int is_new_bgp = 0;
+
+	inst_type = BGP_INSTANCE_TYPE_DEFAULT;
+
 	switch (args->event) {
 	case NB_EV_VALIDATE:
 	case NB_EV_PREPARE:
 	case NB_EV_ABORT:
+		return NB_OK;
 	case NB_EV_APPLY:
-		/* TODO: implement me. */
+		vrf_dnode = yang_dnode_get_parent(args->dnode,
+						  "control-plane-protocol");
+		vrf_name = yang_dnode_get_string(vrf_dnode, "./vrf");
+
+		if (strmatch(vrf_name, VRF_DEFAULT_NAME)) {
+			name = NULL;
+		} else {
+			name = vrf_name;
+			inst_type = BGP_INSTANCE_TYPE_VRF;
+		}
+
+		as = yang_dnode_get_uint32(args->dnode, "./local-as");
+
+		is_view_inst = yang_dnode_get_bool(args->dnode,
+						   "./instance-type-view");
+		if (is_view_inst)
+			inst_type = BGP_INSTANCE_TYPE_VIEW;
+
+		if (inst_type == BGP_INSTANCE_TYPE_DEFAULT)
+			is_new_bgp = (bgp_lookup(as, name) == NULL);
+
+		ret = bgp_get_vty(&bgp, &as, name, inst_type);
```

Comment:
Ideally `bgp_get()` needs to be modified to fetch default values from the YANG model instead of using internal constants for that purpose (e.g. `bgp->default_keepalive = BGP_DEFAULT_KEEPALIVE`). That way we won't need to change the code if we ever want to change default values in the future. Having to maintain internal constants in sync with YANG modules is a burden we can avoid.

Reply:
@Spantik I agree with you that renaming that function would be a good idea, but my comment was about how re retrieve default values from the YANG module. More details here: https://github.com/opensourcerouting/frr/wiki/Retrofitting-Configuration-Commands#default-values